// @mui/x-date-pickers/legacy/internals/hooks/useField/useFieldCharacterEditing.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import * as React from 'react';
import useEventCallback from '@mui/utils/useEventCallback';
import { useUtils } from '../useUtils';
import { applyMeridiemChange, applyWeekDayChange, changeSectionValueFormat, cleanTrailingZeroInNumericSectionValue, doesSectionHaveTrailingZeros, getDateSectionConfigFromFormatToken, getDateSectionGetterAndSetter, getDaysInWeekStr, isFourDigitYearFormat } from './useField.utils';
var QUERY_LIFE_DURATION_MS = 5000;
var isQueryResponseWithoutValue = function isQueryResponseWithoutValue(response) {
  return response.saveQuery != null;
};

/**
 * Update the active section value when the user pressed a key that is not a navigation key (arrow key for example).
 * This hook has two main editing behaviors
 *
 * 1. The numeric editing when the user presses a digit
 * 2. The letter editing when the user presses another key
 */
export var useFieldCharacterEditing = function useFieldCharacterEditing(_ref) {
  var sections = _ref.sections,
    updateSectionValue = _ref.updateSectionValue;
  var utils = useUtils();
  var _React$useState = React.useState(null),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    query = _React$useState2[0],
    setQuery = _React$useState2[1];
  var resetQuery = useEventCallback(function () {
    return setQuery(null);
  });
  React.useEffect(function () {
    var _sections$query$secti;
    if (query != null && ((_sections$query$secti = sections[query.sectionIndex]) == null ? void 0 : _sections$query$secti.dateSectionName) !== query.dateSectionName) {
      resetQuery();
    }
  }, [sections, query, resetQuery]);
  React.useEffect(function () {
    if (query != null) {
      var timeout = setTimeout(function () {
        return resetQuery();
      }, QUERY_LIFE_DURATION_MS);
      return function () {
        window.clearTimeout(timeout);
      };
    }
    return function () {};
  }, [query, resetQuery]);
  var applyQuery = function applyQuery(_ref2, getFirstSectionValueMatchingWithQuery, isValidQueryValue) {
    var keyPressed = _ref2.keyPressed,
      sectionIndex = _ref2.sectionIndex;
    var cleanKeyPressed = keyPressed.toLowerCase();
    var activeSection = sections[sectionIndex];

    // The current query targets the section being editing
    // We can try to concatenated value
    if (query != null && (!isValidQueryValue || isValidQueryValue(query.value)) && query.sectionIndex === sectionIndex) {
      var concatenatedQueryValue = "".concat(query.value).concat(cleanKeyPressed);
      var _queryResponse = getFirstSectionValueMatchingWithQuery(concatenatedQueryValue, activeSection);
      if (!isQueryResponseWithoutValue(_queryResponse)) {
        setQuery({
          sectionIndex: sectionIndex,
          value: concatenatedQueryValue,
          dateSectionName: activeSection.dateSectionName
        });
        return _queryResponse;
      }
    }
    var queryResponse = getFirstSectionValueMatchingWithQuery(cleanKeyPressed, activeSection);
    if (isQueryResponseWithoutValue(queryResponse) && !queryResponse.saveQuery) {
      resetQuery();
      return null;
    }
    setQuery({
      sectionIndex: sectionIndex,
      value: cleanKeyPressed,
      dateSectionName: activeSection.dateSectionName
    });
    if (isQueryResponseWithoutValue(queryResponse)) {
      return null;
    }
    return queryResponse;
  };
  var applyLetterEditing = function applyLetterEditing(params) {
    var findMatchingOptions = function findMatchingOptions(format, options, queryValue) {
      var matchingValues = options.filter(function (option) {
        return option.toLowerCase().startsWith(queryValue);
      });
      if (matchingValues.length === 0) {
        return {
          saveQuery: false
        };
      }
      return {
        sectionValue: matchingValues[0],
        shouldGoToNextSection: matchingValues.length === 1
      };
    };
    var testQueryOnFormatAndFallbackFormat = function testQueryOnFormatAndFallbackFormat(queryValue, activeSection, getOptions, fallbackFormat, formatFallbackValue) {
      if (activeSection.contentType === 'letter') {
        return findMatchingOptions(activeSection.formatValue, getOptions(activeSection.formatValue), queryValue);
      }

      // When editing a digit-format month / weekDay and the user presses a letter,
      // We can support the letter editing by using the letter-format month / weekDay and re-formatting the result.
      // We just have to make sure that the default month / weekDay format is a letter format,
      if (fallbackFormat && formatFallbackValue != null && getDateSectionConfigFromFormatToken(utils, fallbackFormat).contentType === 'letter') {
        var _fallbackOptions = getOptions(fallbackFormat);
        var response = findMatchingOptions(fallbackFormat, _fallbackOptions, queryValue);
        if (isQueryResponseWithoutValue(response)) {
          return {
            saveQuery: false
          };
        }
        return _extends({}, response, {
          sectionValue: formatFallbackValue(response.sectionValue, _fallbackOptions)
        });
      }
      return {
        saveQuery: false
      };
    };
    var getFirstSectionValueMatchingWithQuery = function getFirstSectionValueMatchingWithQuery(queryValue, activeSection) {
      switch (activeSection.dateSectionName) {
        case 'month':
          {
            var getOptions = function getOptions(format) {
              return utils.getMonthArray(utils.date()).map(function (month) {
                return utils.formatByString(month, format);
              });
            };
            var formatFallbackValue = function formatFallbackValue(fallbackValue) {
              return changeSectionValueFormat(utils, fallbackValue, utils.formats.month, activeSection.formatValue);
            };
            return testQueryOnFormatAndFallbackFormat(queryValue, activeSection, getOptions, utils.formats.month, formatFallbackValue);
          }
        case 'weekDay':
          {
            var _getOptions = function _getOptions(format) {
              return getDaysInWeekStr(utils, format);
            };
            var _formatFallbackValue = function _formatFallbackValue(fallbackValue, fallbackOptions) {
              return fallbackOptions.indexOf(fallbackValue).toString();
            };
            return testQueryOnFormatAndFallbackFormat(queryValue, activeSection, _getOptions, utils.formats.weekday, _formatFallbackValue);
          }
        case 'meridiem':
          {
            var _getOptions2 = function _getOptions2(format) {
              var now = utils.date();
              return [utils.endOfDay(now), utils.startOfDay(now)].map(function (date) {
                return utils.formatByString(date, format);
              });
            };
            return testQueryOnFormatAndFallbackFormat(queryValue, activeSection, _getOptions2);
          }
        default:
          {
            return {
              saveQuery: false
            };
          }
      }
    };
    return applyQuery(params, getFirstSectionValueMatchingWithQuery);
  };
  var applyNumericEditing = function applyNumericEditing(params, sectionsValueBoundaries, activeDate) {
    var getNewSectionValue = function getNewSectionValue(queryValue, dateSectionName, format, hasTrailingZeroes, contentType) {
      var queryValueNumber = Number("".concat(queryValue));
      var sectionBoundaries = sectionsValueBoundaries[dateSectionName]({
        currentDate: activeDate,
        format: format,
        contentType: contentType
      });
      if (queryValueNumber > sectionBoundaries.maximum) {
        return {
          saveQuery: false
        };
      }

      // If the user types `0` on a month section,
      // It is below the minimum, but we want to store the `0` in the query,
      // So that when he pressed `1`, it will store `01` and move to the next section.
      if (queryValueNumber < sectionBoundaries.minimum) {
        return {
          saveQuery: true
        };
      }
      var shouldGoToNextSection = Number("".concat(queryValue, "0")) > sectionBoundaries.maximum || queryValue.length === sectionBoundaries.maximum.toString().length;

      // queryValue without trailing `0` (`01` => `1`)
      var newSectionValue = queryValueNumber.toString();
      if (hasTrailingZeroes) {
        newSectionValue = cleanTrailingZeroInNumericSectionValue(utils, format, newSectionValue);
      }
      return {
        sectionValue: newSectionValue,
        shouldGoToNextSection: shouldGoToNextSection
      };
    };
    var getFirstSectionValueMatchingWithQuery = function getFirstSectionValueMatchingWithQuery(queryValue, activeSection) {
      if (activeSection.contentType === 'digit') {
        return getNewSectionValue(queryValue, activeSection.dateSectionName, activeSection.formatValue, activeSection.hasTrailingZeroes, activeSection.contentType);
      }

      // When editing a letter-format month and the user presses a digit,
      // We can support the numeric editing by using the digit-format month and re-formatting the result.
      if (activeSection.dateSectionName === 'month') {
        var response = getNewSectionValue(queryValue, activeSection.dateSectionName, 'MM', doesSectionHaveTrailingZeros(utils, 'digit', 'month', 'MM'), 'digit');
        if (isQueryResponseWithoutValue(response)) {
          return response;
        }
        var formattedValue = changeSectionValueFormat(utils, response.sectionValue, 'MM', activeSection.formatValue);
        return _extends({}, response, {
          sectionValue: formattedValue
        });
      }

      // When editing a letter-format weekDay and the user presses a digit,
      // We can support the numeric editing by returning the nth day in the week day array.
      if (activeSection.dateSectionName === 'weekDay') {
        var _response = getNewSectionValue(queryValue, activeSection.dateSectionName, activeSection.formatValue, activeSection.hasTrailingZeroes, activeSection.contentType);
        if (isQueryResponseWithoutValue(_response)) {
          return _response;
        }
        var _formattedValue = getDaysInWeekStr(utils, activeSection.formatValue)[Number(_response.sectionValue) - 1];
        return _extends({}, _response, {
          sectionValue: _formattedValue
        });
      }
      return {
        saveQuery: false
      };
    };
    return applyQuery(params, getFirstSectionValueMatchingWithQuery, function (queryValue) {
      return !Number.isNaN(Number(queryValue));
    });
  };
  var applyCharacterEditing = useEventCallback(function (params) {
    var activeSection = sections[params.sectionIndex];
    var isNumericEditing = !Number.isNaN(Number(params.keyPressed));
    var getNewSectionValue = isNumericEditing ? applyNumericEditing : applyLetterEditing;
    updateSectionValue({
      activeSection: activeSection,
      setSectionValueOnDate: function setSectionValueOnDate(activeDate, sectionsValueBoundaries) {
        var response = getNewSectionValue(params, sectionsValueBoundaries, activeDate);
        if (response == null) {
          return null;
        }
        if (activeSection.dateSectionName === 'meridiem') {
          var _newDate = applyMeridiemChange(utils, activeDate, response.sectionValue);
          return {
            date: _newDate,
            shouldGoToNextSection: true
          };
        }
        if (activeSection.dateSectionName === 'weekDay') {
          var _newDate2 = applyWeekDayChange(utils, activeDate, activeSection.formatValue, response.sectionValue);
          return {
            date: _newDate2,
            shouldGoToNextSection: response.shouldGoToNextSection
          };
        }
        var _getDateSectionGetter = getDateSectionGetterAndSetter(utils, activeSection.dateSectionName),
          getter = _getDateSectionGetter.getter,
          setter = _getDateSectionGetter.setter;
        var newSectionValue;
        // We can't parse the day on the current date, otherwise we might try to parse `31` on a 30-days month.
        // So we take for granted that for days, the digit rendered is always 1-indexed, just like the digit stored in the date.
        if (activeSection.contentType === 'digit' && activeSection.dateSectionName === 'day') {
          newSectionValue = Number(response.sectionValue);
        }
        // Day.js (and maybe other adapters) do not support correctly the parsing of standalone years.
        // To get the proper value
        // - if the format is a 4-digit year, we just use the section value.
        // - if the format is a 2-digit year, we add a trailing zero if needed and then parse the date.
        else if (activeSection.contentType === 'digit' && activeSection.dateSectionName === 'year') {
          if (isFourDigitYearFormat(utils, activeSection.formatValue)) {
            newSectionValue = Number(response.sectionValue);
          } else {
            var cleanValue = response.sectionValue.length === 1 ? "0".concat(response.sectionValue) : response.sectionValue;
            var sectionDate = utils.parse(cleanValue, activeSection.formatValue);
            newSectionValue = getter(sectionDate);
          }
        } else {
          // The month is stored as 0-indexed in the date (0 = January, 1 = February, ...).
          // But it is often rendered as 1-indexed in the input (1 = January, 2 = February, ...).
          // This parsing makes sure that we store the digit according to the date index and not the input index.
          var _sectionDate = utils.parse(response.sectionValue, activeSection.formatValue);
          newSectionValue = getter(_sectionDate);
        }
        var newDate = setter(activeDate, newSectionValue);
        return {
          date: newDate,
          shouldGoToNextSection: response.shouldGoToNextSection
        };
      },
      setSectionValueOnSections: function setSectionValueOnSections(sectionsValueBoundaries) {
        return getNewSectionValue(params, sectionsValueBoundaries, null);
      }
    });
  });
  return {
    applyCharacterEditing: applyCharacterEditing,
    resetCharacterQuery: resetQuery
  };
};
