// @mui/x-date-pickers/node/AdapterMomentJalaali/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterMomentJalaali = void 0;
var _jalaali = _interopRequireDefault(require("@date-io/jalaali"));
/* eslint-disable class-methods-use-this */

// From https://momentjs.com/docs/#/displaying/format/
const formatTokenMap = {
  // Month
  jM: 'month',
  jMo: 'month',
  jMM: 'month',
  jMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  jMMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  // Day of Month
  jD: 'day',
  jDo: 'day',
  jDD: 'day',
  // Year
  jY: 'year',
  jYY: 'year',
  jYYYY: 'year',
  jYYYYYY: 'year',
  // AM / PM
  A: 'meridiem',
  a: 'meridiem',
  // Hour
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  k: 'hours',
  kk: 'hours',
  // Minute
  m: 'minutes',
  mm: 'minutes',
  // Second
  s: 'seconds',
  ss: 'seconds'
};
class AdapterMomentJalaali extends _jalaali.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: '[',
      end: ']'
    };
    this.expandFormat = format => {
      // @see https://github.com/moment/moment/blob/develop/src/lib/format/format.js#L6
      const localFormattingTokens = /(\[[^[]*\])|(\\)?(LTS|LT|LL?L?L?|l{1,4})|./g;
      return format.match(localFormattingTokens).map(token => {
        const firstCharacter = token[0];
        if (firstCharacter === 'L' || firstCharacter === ';') {
          return this.moment.localeData(this.getCurrentLocaleCode()).longDateFormat(token);
        }
        return token;
      }).join('').replace('dd', 'jDD'); // Fix for https://github.com/dmtrKovalenko/date-io/pull/632;
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/a/gi, '(a|p)m').replace('jY', 'Y').replace('jM', 'M').replace('jD', 'D').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return date.jWeek();
    };
    this.addYears = (date, count) => {
      return count < 0 ? date.clone().subtract(Math.abs(count), 'jYear') : date.clone().add(count, 'jYear');
    };
    this.addMonths = (date, count) => {
      return count < 0 ? date.clone().subtract(Math.abs(count), 'jMonth') : date.clone().add(count, 'jMonth');
    };
    this.isValid = value => {
      // We can't to `this.moment(value)` because moment-jalaali looses the invalidity information when creating a new moment object from an existing one
      if (!this.moment.isMoment(value)) {
        return false;
      }
      return value.isValid(value);
    };
  }
}
exports.AdapterMomentJalaali = AdapterMomentJalaali;
