// @mui/x-date-pickers/node/AdapterDayjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterDayjs = void 0;
var _dayjs = _interopRequireDefault(require("dayjs"));
var _weekOfYear = _interopRequireDefault(require("dayjs/plugin/weekOfYear"));
var _dayjs2 = _interopRequireDefault(require("@date-io/dayjs"));
var _warning = require("../internals/utils/warning");
/* eslint-disable class-methods-use-this */

const localeNotFoundWarning = (0, _warning.buildWarning)(['Your locale has not been found.', 'Either the locale key is not a supported one. Locales supported by dayjs are available here: https://github.com/iamkun/dayjs/tree/dev/src/locale', "Or you forget to import the locale with `require('dayjs/locale/{localeUsed}')`", 'fallback on English locale']);
const formatTokenMap = {
  YY: 'year',
  YYYY: 'year',
  M: 'month',
  MM: 'month',
  MMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  D: 'day',
  DD: 'day',
  d: 'weekDay',
  dd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  ddd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  dddd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  m: 'minutes',
  mm: 'minutes',
  s: 'seconds',
  ss: 'seconds',
  A: 'meridiem',
  a: 'meridiem'
};
class AdapterDayjs extends _dayjs2.default {
  constructor(options) {
    super(options);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: '[',
      end: ']'
    };
    this.getLocaleFormats = () => {
      const locales = this.rawDayJsInstance.Ls ?? _dayjs.default.Ls;
      const locale = this.locale || 'en';
      let localeObject = locales[locale];
      if (localeObject === undefined) {
        localeNotFoundWarning();
        localeObject = locales.en;
      }
      return localeObject.formats;
    };
    this.is12HourCycleInCurrentLocale = () => {
      /* istanbul ignore next */
      return /A|a/.test(this.getLocaleFormats().LT || '');
    };
    this.expandFormat = format => {
      const localeFormats = this.getLocaleFormats();

      // @see https://github.com/iamkun/dayjs/blob/dev/src/plugin/localizedFormat/index.js
      const t = formatBis => formatBis.replace(/(\[[^\]]+])|(MMMM|MM|DD|dddd)/g, (_, a, b) => a || b.slice(1));
      return format.replace(/(\[[^\]]+])|(LTS?|l{1,4}|L{1,4})/g, (_, a, b) => {
        const B = b && b.toUpperCase();
        return a || localeFormats[b] || t(localeFormats[B]);
      });
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/a/gi, '(a|p)m').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return date.week();
    };
    _dayjs.default.extend(_weekOfYear.default);
  }
}
exports.AdapterDayjs = AdapterDayjs;
