// @mui/x-date-pickers/node/AdapterLuxon/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterLuxon = void 0;
var _luxon = require("luxon");
var _luxon2 = _interopRequireDefault(require("@date-io/luxon"));
/* eslint-disable class-methods-use-this */

const formatTokenMap = {
  s: 'seconds',
  ss: 'seconds',
  m: 'minutes',
  mm: 'minutes',
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  a: 'meridiem',
  d: 'day',
  dd: 'day',
  c: 'weekDay',
  ccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  cccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  ccccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  E: 'weekDay',
  EEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EEEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EEEEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  L: 'month',
  LL: 'month',
  LLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLLLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  M: 'month',
  MM: 'month',
  MMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  y: 'year',
  yy: 'year',
  yyyy: 'year'
};
class AdapterLuxon extends _luxon2.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: "'",
      end: "'"
    };
    this.expandFormat = format => {
      if (!_luxon.DateTime.expandFormat) {
        throw Error('Your luxon version does not support `expandFormat`. Consider upgrading it to v3.0.2');
      }
      // The returned format can contain `yyyyy` which means year between 4 and 6 digits.
      // This value is supported by luxon parser but not luxon formatter.
      // To avoid conflicts, we replace it by 4 digits which is enough for most use-cases.
      return _luxon.DateTime.expandFormat(format, {
        locale: this.locale
      }).replace('yyyyy', 'yyyy');
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/(a)/g, '(a|p)m').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return date.weekNumber;
    };
  }
}
exports.AdapterLuxon = AdapterLuxon;
