// @mui/x-date-pickers/node/AdapterMoment/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterMoment = void 0;
var _moment = _interopRequireDefault(require("@date-io/moment"));
/* eslint-disable class-methods-use-this */

// From https://momentjs.com/docs/#/displaying/format/
const formatTokenMap = {
  // Month
  M: 'month',
  Mo: 'month',
  MM: 'month',
  MMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  // Day of the month
  D: 'day',
  Do: 'day',
  DD: 'day',
  // Day of the week
  E: 'weekDay',
  e: 'weekDay',
  d: 'weekDay',
  dd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  ddd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  dddd: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  // Year
  Y: 'year',
  YY: 'year',
  YYYY: 'year',
  YYYYYY: 'year',
  // AM / PM
  A: 'meridiem',
  a: 'meridiem',
  // Hour
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  k: 'hours',
  kk: 'hours',
  // Minute
  m: 'minutes',
  mm: 'minutes',
  // Second
  s: 'seconds',
  ss: 'seconds'
};
class AdapterMoment extends _moment.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: '[',
      end: ']'
    };
    this.expandFormat = format => {
      // @see https://github.com/moment/moment/blob/develop/src/lib/format/format.js#L6
      const localFormattingTokens = /(\[[^[]*\])|(\\)?(LTS|LT|LL?L?L?|l{1,4})|./g;
      return format.match(localFormattingTokens).map(token => {
        const firstCharacter = token[0];
        if (firstCharacter === 'L' || firstCharacter === ';') {
          return this.moment.localeData(this.getCurrentLocaleCode()).longDateFormat(token);
        }
        return token;
      }).join('');
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/a/gi, '(a|p)m').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return date.week();
    };
  }
}
exports.AdapterMoment = AdapterMoment;
