// @mui/x-date-pickers/node/AdapterMomentHijri/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterMomentHijri = void 0;
var _hijri = _interopRequireDefault(require("@date-io/hijri"));
/* eslint-disable class-methods-use-this */

// From https://momentjs.com/docs/#/displaying/format/
const formatTokenMap = {
  // Month
  iM: 'month',
  iMo: 'month',
  iMM: 'month',
  iMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  iMMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  // Day of Month
  iD: 'day',
  iDo: 'day',
  iDD: 'day',
  // Year
  iY: 'year',
  iYY: 'year',
  iYYYY: 'year',
  iYYYYYY: 'year',
  // AM / PM
  A: 'meridiem',
  a: 'meridiem',
  // Hour
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  k: 'hours',
  kk: 'hours',
  // Minute
  m: 'minutes',
  mm: 'minutes',
  // Second
  s: 'seconds',
  ss: 'seconds'
};
class AdapterMomentHijri extends _hijri.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: '[',
      end: ']'
    };
    this.expandFormat = format => {
      // @see https://github.com/moment/moment/blob/develop/src/lib/format/format.js#L6
      const localFormattingTokens = /(\[[^[]*\])|(\\)?(LTS|LT|LL?L?L?|l{1,4})|./g;
      return format.match(localFormattingTokens).map(token => {
        const firstCharacter = token[0];
        if (firstCharacter === 'L' || firstCharacter === ';') {
          return this.moment.localeData(this.getCurrentLocaleCode()).longDateFormat(token);
        }
        return token;
      }).join('').replace('dd', 'iDD'); // Fix for https://github.com/dmtrKovalenko/date-io/pull/632
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/a/gi, '(a|p)m').replace('iY', 'Y').replace('iM', 'M').replace('iD', 'D').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return date.iWeek();
    };
  }
}
exports.AdapterMomentHijri = AdapterMomentHijri;
