// @mui/x-date-pickers/node/AdapterDateFnsJalali/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterDateFnsJalali = void 0;
var _dateFnsJalali = _interopRequireDefault(require("@date-io/date-fns-jalali"));
var _faIR = _interopRequireDefault(require("date-fns-jalali/locale/fa-IR"));
var _getWeek = _interopRequireDefault(require("date-fns-jalali/getWeek"));
var _longFormatters = _interopRequireDefault(require("date-fns-jalali/_lib/format/longFormatters"));
// @ts-ignore

const formatTokenMap = {
  y: 'year',
  yy: 'year',
  yyy: 'year',
  yyyy: 'year',
  M: 'month',
  MM: 'month',
  MMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  d: 'day',
  dd: 'day',
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  mm: 'minutes',
  ss: 'seconds',
  a: 'meridiem',
  aa: 'meridiem',
  aaa: 'meridiem'
};
class AdapterDateFnsJalali extends _dateFnsJalali.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: "'",
      end: "'"
    };
    this.expandFormat = format => {
      // @see https://github.com/date-fns/date-fns/blob/master/src/format/index.js#L31
      const longFormatRegexp = /P+p+|P+|p+|''|'(''|[^'])+('|$)|./g;
      const locale = this.locale ?? _faIR.default;
      return format.match(longFormatRegexp).map(token => {
        const firstCharacter = token[0];
        if (firstCharacter === 'p' || firstCharacter === 'P') {
          const longFormatter = _longFormatters.default[firstCharacter];
          return longFormatter(token, locale.formatLong, {});
        }
        return token;
      }).join('');
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/(aaa|aa|a)/g, '(a|p)m').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return (0, _getWeek.default)(date, {
        locale: this.locale
      });
    };
  }
}
exports.AdapterDateFnsJalali = AdapterDateFnsJalali;
