// @mui/x-date-pickers/node/AdapterDateFns/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdapterDateFns = void 0;
var _dateFns = _interopRequireDefault(require("@date-io/date-fns"));
var _enUS = _interopRequireDefault(require("date-fns/locale/en-US"));
var _longFormatters = _interopRequireDefault(require("date-fns/_lib/format/longFormatters"));
var _getWeek = _interopRequireDefault(require("date-fns/getWeek"));
// @ts-ignore

const formatTokenMap = {
  y: 'year',
  yy: 'year',
  yyy: 'year',
  yyyy: 'year',
  M: 'month',
  MM: 'month',
  MMMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  MMM: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  LLLL: {
    sectionName: 'month',
    contentType: 'letter'
  },
  E: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EEEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  EEEEE: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  i: 'weekDay',
  ii: 'weekDay',
  iii: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  iiii: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  e: 'weekDay',
  ee: 'weekDay',
  eee: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  eeee: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  eeeee: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  eeeeee: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  c: 'weekDay',
  cc: 'weekDay',
  ccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  cccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  ccccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  cccccc: {
    sectionName: 'weekDay',
    contentType: 'letter'
  },
  d: 'day',
  dd: 'day',
  H: 'hours',
  HH: 'hours',
  h: 'hours',
  hh: 'hours',
  mm: 'minutes',
  ss: 'seconds',
  a: 'meridiem',
  aa: 'meridiem',
  aaa: 'meridiem'
};
class AdapterDateFns extends _dateFns.default {
  constructor(...args) {
    super(...args);
    this.isMUIAdapter = true;
    this.formatTokenMap = formatTokenMap;
    this.escapedCharacters = {
      start: "'",
      end: "'"
    };
    this.expandFormat = format => {
      const longFormatRegexp = /P+p+|P+|p+|''|'(''|[^'])+('|$)|./g;

      // @see https://github.com/date-fns/date-fns/blob/master/src/format/index.js#L31
      return format.match(longFormatRegexp).map(token => {
        const firstCharacter = token[0];
        if (firstCharacter === 'p' || firstCharacter === 'P') {
          const longFormatter = _longFormatters.default[firstCharacter];
          const locale = this.locale || _enUS.default;
          return longFormatter(token, locale.formatLong, {});
        }
        return token;
      }).join('');
    };
    this.getFormatHelperText = format => {
      return this.expandFormat(format).replace(/(aaa|aa|a)/g, '(a|p)m').toLocaleLowerCase();
    };
    this.getWeekNumber = date => {
      return (0, _getWeek.default)(date, {
        locale: this.locale
      });
    };
  }
}
exports.AdapterDateFns = AdapterDateFns;
