// @mui/x-date-pickers/legacy/tests/describeAdapters/describeAdapters.js
import _extends from "@babel/runtime/helpers/esm/extends";
import moment from 'moment';
import createDescribe from '@mui/monorepo/test/utils/createDescribe';
import { buildFieldInteractions, createPickerRenderer } from 'test/utils/pickers-utils';
var ADAPTERS = ['dayjs', 'date-fns', 'luxon', 'moment'];
function innerDescribeAdapters(title, FieldComponent, testRunner) {
  ADAPTERS.forEach(function (adapterName) {
    describe("".concat(title, " - adapter: ").concat(adapterName), function () {
      if (adapterName === 'moment') {
        moment.locale('en');
      }
      var pickerRendererResponse = createPickerRenderer({
        adapterName: adapterName,
        clock: 'fake',
        clockConfig: new Date(2022, 5, 15)
      });
      var fieldInteractions = buildFieldInteractions({
        clock: pickerRendererResponse.clock,
        render: pickerRendererResponse.render,
        Component: FieldComponent
      });
      testRunner(_extends({
        adapterName: adapterName
      }, pickerRendererResponse, fieldInteractions));
    });
  });
}
export var describeAdapters = createDescribe('Adapter specific tests', innerDescribeAdapters);
